actors/evm/src/interpreter/precompiles.rs

```diff
@@ -93,7 +103,75 @@ fn read_right_pad<'a>(input: impl Into<Cow<'a, [u8]>>, len: usize) -> Cow<'a, [u
     input
 }
 
-// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L165
+fn todo<RT: Runtime>(_: &RT, _: &[u8]) -> PrecompileResult {
+    todo!()
+}
+
+fn get_actor_code_cid<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    let id = EthAddress(input[..20].try_into().unwrap()).as_id();
+    if id.is_none() {
+        return Err(PrecompileError::InvalidInput);
+    }
+    Ok(rt.get_actor_code_cid(&id.unwrap()).unwrap_or_default().to_bytes())
+}
+
+fn get_randomness<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    // 1 + 7 (reserved) + 8 + 8 + 8 = 32
+    let word = read_right_pad(input, 32);
+
+    #[derive(num_derive::FromPrimitive)]
+    #[repr(u8)]
+    enum RandomnessType {
+        Chain = 0,
+        Beacon = 1,
+    }
+
+    let randomness_type = RandomnessType::from_u8(word[0]);
+    // 7 bytes reserved
+    let personalization = i64::from_le_bytes(word[8..16].try_into().unwrap());
+    let rand_epoch = i64::from_le_bytes(word[16..24].try_into().unwrap());
+    let entropy_len = u64::from_le_bytes(word[24..32].try_into().unwrap());
+
+    let entropy = read_right_pad(&input[32..], entropy_len as usize);
+
+    let randomness = match randomness_type {
+        Some(RandomnessType::Chain) => rt
+            .get_randomness_from_tickets(personalization, rand_epoch, &entropy)
+            .map(|a| a.to_vec()),
+        Some(RandomnessType::Beacon) => {
+            rt.get_randomness_from_beacon(personalization, rand_epoch, &entropy).map(|a| a.to_vec())
+        }
+        None => Ok(Vec::new()),
+    };
+
+    randomness.map_err(|_| PrecompileError::InvalidInput)
+}
+
+// looks up the other address of an ID address, returns empty array if not found, or the FIL encoded address
+// TODO what is predictable here for lookup address: only f2 and f4? f4 addresses arent always predictable...
+// TODO should we assume raw actor ID or EVM style FIL ID address?
+fn lookup_address<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    let id = EthAddress(input[..20].try_into().unwrap()).as_id();
+    if id.is_none() {
+        return Err(PrecompileError::InvalidInput);
+    }
+    let address = rt.lookup_address(id.unwrap());
+    let ab = match address {
+        Some(a) => a.to_bytes(),
+        None => Vec::new(),
+    };
+    Ok(ab)
+}
+
+// REMOVEME i'd like to be able to return with 0, but 0 id address is system, so ima just use empty vec :\
```

Comment:
Yeah, that's annoying. We _can_ just return 0xfffffffff..., but that's not great either.